src/foam/nanos/test/SmartTest.java

```diff
@@ -0,0 +1,145 @@
+package foam.nanos.test;
+
+import foam.core.X;
+import foam.core.EmptyX;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.lang.System.err;
+
+public class SmartTest extends Test {
+  public static final String TEST_METHOD_SUFFIX = "Test";
+  public static final String CLEANUP_METHOD_SUFFIX = "Cleanup";
+  public static final String SETUP_METHOD_SUFFIX = "Setup";
+
+  private Method globalSetupMethod;
+  private Method globalCleanupMethod;
+  private Method preTestMethod;
+  private Method postTestMethod;
+  private final Map<String, TestChain> testChains = new HashMap<>();
+
+  protected X x_ = EmptyX.instance();
+
+  public X getX() {
+    return x_;
+  }
+
+  public void setX(X x) {
+    x_ = x;
+  }
+
+  @Override
+  public void runTest(X x) {
+    setX(x);
+    prepareTestChains();
+    runTestChains();
+  }
+
+  private void runTestChains() {
+    if ( invokeMethod(globalSetupMethod) ) {
+      for ( String test: testChains.keySet() ) {
+        TestChain testChain = testChains.get(test);
+        runTestChain(testChain);
+      }
+      invokeMethod(globalCleanupMethod);
+    }
+  }
+
+  private void runTestChain(TestChain testChain) {
+
+    // don't run when the test Chain if doesn't have a test method
+    if ( testChain.test == null ) {
+      return;
+    }
+    if ( invokeMethod(preTestMethod) ) {
+      if ( invokeMethod(testChain.setup) )  {
```

Comment:
I think the setup method should return a Context for use by test and cleanup.

Reply:
Wouldn't this be the responsibility of the person writing the tests? Like saving the context created in the setup as a class variable, and using that in the cleanup after the test is ran?

Reply:
The setup for each test would be responsible for building a sub-context. If this were stored in an instance variable then you have a lot of negative potential for tests not resetting the context or inadvertently running in subcontexts of other tests. When the setup and the test are defined in the same method subcontexts can just be local variables which can't escape the test but now that setup and test are different methods we need a method for passing along the sub-context in an organized way which isn't going to mess things up. Also, it occurs to me that many tests may want to create multiple sub-contexts, in which case having a separate setup method is going to be problematic.